DBZ-9329 - Memory Leak Fix for Debezium MySQL Connector

yashisrivastava-confluent opened the pull request:
Fixes debezium/dbz#1383

Fix memory leak in binlog-based connectors during snapshot interruptions

This PR resolves coordinator thread memory leaks that occur when snapshot operations are interrupted, causing threads to remain blocked indefinitely in ChangeEventQueue operations.

**Key Changes:**
- Added thread-safe shutdown mechanism to ChangeEventQueue with graceful interruption
- Created BinlogSourceTask and BinlogChangeEventSourceFactory base classes for code reuse
- Updated MySQL and MariaDB connectors to use centralized shutdown logic
- Added comprehensive test coverage (10 test scenarios) for shutdown edge cases

**Impact:** Prevents resource exhaustion in production deployments by ensuring coordinator threads can exit cleanly when connectors are stopped during snapshot operations.

yashisrivastava-confluent committed:
DBZ-9329 Memory Leak Fix for Debezium MySQL Connector

This commit implements a queue shutdown mechanism to gracefully interrupt
blocked threads and prevent memory leaks during task shutdown scenarios.

Changes:
1. ChangeEventQueue.java - Added shutdown() method and shutdown checks in doEnqueue() and flushBuffer()
2. MySqlConnectorTask.java - Added early queue shutdown in doStop() method
3. MySqlChangeEventSourceFactory.java - Enhanced modifyAndFlushLastRecord() with graceful shutdown handling

Fixes coordinator threads stuck indefinitely in queue.flushBuffer() calls
and prevents memory accumulation when snapshot operations are interrupted.

yashisrivastava-confluent committed:
DBZ-9329 Improve modifyAndFlushLastRecord with robust thread interruption handling

Enhanced the modifyAndFlushLastRecord method with:
- Early thread interruption detection before attempting flush
- Robust finally block to ensure queue.disableBuffering() always runs
- Better error handling for AssertionError during cleanup
- More informative logging for debugging
- Graceful handling of shutdown timing edge cases

This provides more reliable cleanup and better thread safety during
connector shutdown scenarios.

yashisrivastava-confluent committed:
DBZ-9329 Fix thread safety and cleanup issues in ChangeEventQueue

Fixed the following issues:
1. Thread safety: Move 'running = false' assignment inside the lock in shutdown() method
   to prevent race conditions with other threads checking the running flag
2. Remove duplicate isBuffered() method, keeping only isBuffering() with proper documentation

This ensures proper synchronization during shutdown and cleaner API.

yashisrivastava-confluent committed:
DBZ-9329 Add comprehensive test suite for ChangeEventQueue shutdown mechanism

Added ChangeEventQueueShutdownTest.java with the following test cases:
1. shouldUnblockThreadsOnShutdown - Tests single thread unblocking
2. shouldUnblockMultipleThreads - Tests multiple threads unblocking
3. shouldHandleMultipleShutdownCalls - Tests idempotent shutdown behavior
4. shouldUnblockFlushBufferOnShutdown - Tests flushBuffer interruption
5. shouldReflectRunningState - Tests isRunning() state management

These tests validate the memory leak fix by ensuring that:
- Blocked enqueue operations are interrupted immediately on shutdown
- Multiple threads are all unblocked simultaneously
- Shutdown is safe to call multiple times
- Buffer flush operations are also properly interrupted
- Queue state is correctly tracked

This provides comprehensive coverage for the coordinator thread memory leak fix.

Naros commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorTask.java:
```diff
@@ -274,6 +274,12 @@ protected Optional<ErrorHandler> getErrorHandler() {
 
     @Override
     protected void doStop() {
+        // Shut down the queue early to prevent memory leaks from blocked threads
+        if (queue != null) {
+            LOGGER.info("Shutting down queue to prevent thread leaks");
+            queue.shutdown();
+        }
```
Would this also be necessary for MariaDB? If so, it may be useful to create a `doStop()` implementation in the `BinlogSourceTask` class and place this logic there, having both MySQL and MariaDB call their respective super type methods.  What do you think?

Naros commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlChangeEventSourceFactory.java:
```diff
@@ -85,8 +89,35 @@ private void preSnapshot() {
     }
 
     private void modifyAndFlushLastRecord(Function<SourceRecord, SourceRecord> modify) throws InterruptedException {
```
This likely applies to MariaDB. I might suggest we create a `BinlogChangeEventSourceFactory` abstract base class that both MySQL and MariaDB extend from, and put this method in the abstract base class.

Naros reviewed (changes requested):
I've left a few questions around refactoring and applicability.

yashisrivastava-confluent committed:
DBZ-9329 Fix compilation error: Update EventDispatcher to use isBuffering() method

The EventDispatcher was still calling the removed isBuffered() method.
Updated to use the correct isBuffering() method that we kept in ChangeEventQueue.

This fixes the compilation failure in the CI build.

yashisrivastava-confluent committed:
DBZ-9329 Fix memory leak in Debezium MySQL/MariaDB connectors

Implement comprehensive solution to prevent memory accumulation and coordinator
thread leaks when snapshot operations are interrupted.

Core Changes:
- Add shutdown mechanism to ChangeEventQueue with thread-safe interruption
- Implement graceful queue shutdown that unblocks waiting threads
- Add isRunning() method to track queue operational state

Base Class Refactoring:
- Create BinlogSourceTask abstract base class for shared doStop() logic
- Create BinlogChangeEventSourceFactory base class for snapshot cleanup
- Centralize memory leak prevention across MySQL and MariaDB connectors

Connector Updates:
- Update MySqlConnectorTask and MariaDbConnectorTask to extend BinlogSourceTask
- Update factory classes to extend BinlogChangeEventSourceFactory
- Ensure early queue shutdown prevents thread blocking

Testing:
- Add comprehensive ChangeEventQueueShutdownTest with multiple scenarios
- Add BinlogChangeEventSourceFactoryTest for base class functionality
- Test thread interruption, queue shutdown, and edge cases

Code Quality:
- Fix all formatting and checkstyle issues
- Remove unused imports and methods
- Maintain backward compatibility with existing APIs

This fix prevents stuck coordinator threads and memory accumulation during
connector shutdown, particularly when snapshot operations are interrupted.

yashisrivastava-confluent commented:
@Naros I have made the fixes recommended by you, can I get a re-review on the same?

yashisrivastava-confluent committed:
DBZ-9329 Queue shutdown in base class as per review

yashisrivastava-confluent committed:
DBZ-9329 Fix lint and override

Naros reviewed (commented):
Overall, this LGTM. 

The only question I have, and I am on the fence about it, is whether or not we should be setting up every connector to return its _queue_ in the `getChangeEventQueue`. Do you have any preference @jpechane ?

yashisrivastava-confluent committed:
DBZ-9329 Fix lint

jpechane commented on debezium-connector-binlog/src/main/java/io/debezium/connector/binlog/BinlogChangeEventSourceFactory.java:
```diff
@@ -0,0 +1,73 @@
+            try {
+                queue.disableBuffering();
+            }
+            catch (AssertionError e) {
```
I definitely don't like this part of the code. If it is acceptable to have something when buffering is disabled in the buffer then assertion should not fail. Maybe and extra method needs to be added to reflect that fact?

jpechane commented:
> Overall, this LGTM.
> 
> The only question I have, and I am on the fence about it, is whether or not we should be setting up every connector to return its _queue_ in the `getChangeEventQueue`. Do you have any preference @jpechane ?

Do we have any connector that does not use queue? I don;t think so. If it is the case then I believe the current implementation is OK and there should be a follow-up PR refactoring that would pull queue up into `BaseSourceTask`.

jpechane commented on debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:
```diff
@@ -241,15 +244,19 @@ protected void doEnqueue(T record) throws InterruptedException {
-            queue.enqueue(record);
-            // If we pass a positiveLong max.queue.size.in.bytes to enable handling queue size in bytes feature
+            // If not running, throw InterruptedException to allow graceful exit
+            if (!running) {
```
Is `volatile` enough here? What if `running` flag is change between this check and the next `enqueue` call?

jpechane commented on debezium-core/src/main/java/io/debezium/connector/common/BaseSourceTask.java:
```diff
@@ -507,6 +516,18 @@ private void stop(boolean restart) {
+     *
+     * @return the ChangeEventQueue instance or null if not applicable
+     */
+    protected ChangeEventQueue<DataChangeEvent> getChangeEventQueue() {
```
I'd prefer `Optional` here if possible.

jpechane reviewed (commented):
@yashisrivastava-confluent Thanks for the PR. It definitely goes into right direction. Could you please take a look at my comments? Thanks

yashisrivastava-confluent committed:
DBZ-9329 Fix other

yashisrivastava-confluent commented:
@jpechane I have made requested changes, please re-review

jpechane commented on support/checkstyle/src/main/java/io/debezium/checkstyle/ExcludeTestPackages.java:
```diff
@@ -11,9 +11,9 @@
- * Allows disabling some rules for the test suite source. Any violation will be suppressed if it's generated by a source file
- * having {@code /src/test/java} included in its path and if the violation message contains the keyword
- * "[not required for tests]".
+ * Allows disabling some rules for the test suite source. Any violation will be suppressed if it's generated by a source
```
Please don't include change sin `support` module.

jpechane commented on debezium-core/src/main/java/io/debezium/connector/base/ChangeEventQueue.java:
```diff
@@ -212,20 +213,44 @@ record = bufferedEvent.getAndSet(record);
+            // Clear the buffer to prevent memory leaks
+            bufferedEvent.set(null);
+        }
+        buffering = false;
```
I think this should go to the top of the method. Otherwise you can celan the buffer but another thread can fill it again before buffering is disabled.

Naros commented:
Hi @yashisrivastava-confluent are you in a position to rebase this PR, add commit sign-offs and address Jiri's remaining open questions?
